### ISSUE TITLE
Passing events through to connection process

### ISSUE THREAD

**Issue by MorgVerd** [ronf/asyncssh#502]
Hello, I've been trying to make a centeralised SSH server that proxies connections (for central authentication). In [this thread](https://github.com/ronf/asyncssh/issues/468) an example was posted for doing this, however one issue I am facing is that I can't figure out how to also pass through events.

For example, the TerminalSizeChanged event is never passed through to the connection, so it only uses its initial values from when it was initialised. Any tips would be appriciated if anyone's found a solution for this.

**Comment by ronf** [comment]
Yeah - I see what's going on here. I never really thought about the SSHProcess redirect functionality redirecting things like breaks, signals, and terminal size changes. As a result, they're still queued up in the receive buffers in the SSHSession classes. If you had your own sessions, it would be straightforward to implement the callback functions such as terminal_size_changed() on an SSHServerSession and then have that call change_terminal_size() on the upstream SSHClientSession. However, you'd have to do this same sort of them for all the data channels to forward input/output and EOF on each.

It looks like the stream session classes can still read the queued up signals, even with redirect active. So, a workaround for now would be to add a function like:

```python
async def handle_signals(process, upstream_process):
    while not process.stdin.at_eof():
        try:
            await process.stdin.read()
        except asyncssh.TerminalSizeChanged as exc:
            upstream_process.change_terminal_size(exc.width, exc.height,
                                                  exc.pixwidth, exc.pixheight)
```

Then, in your handle_client() function, add the following before the call to wait_closed():

```python
        await asyncio.create_task(handle_signals(process, upstream_process))
```

This will read and forward any terminal size changes. If you needed it, you could follow a similar approach for BreakReceived and SignalReceived.

Ideally, this forwarding would happen automatically when you did a redirect from stdin on an SSHServerProcess to stdin on an SSHClientProcess. I'll look into what it would take to do this.

**Comment by MorgVerd** [comment]
Thanks for the help, the workaround seems to be working fine currently!

**Comment by ronf** [comment]
Great! Note that I did edit the example above slightly this morning. The call to create_task() should NOT have an "await" in front of it. If it does, it'll prevent handle_client() from properly exiting when the upstream process exits.

I'm also making progress on making this forwarding automatic. I just need to add some unit tests. With the change I'm making here, the original example code should properly handle terminal size changes (along with breaks and signals) when redirection is enabled without needing to add the handle_signals() task.

**Comment by ronf** [comment]
If you'd like to give it a try, this new feature is now available as commit f389a67 in the AsyncSSH "develop" branch, and it will be included in the next release.

**Comment by ronf** [comment]
This new feature is now available in AsyncSSH 2.12.0.
